python/tritonserver/_api/wrapper.py

```diff
@@ -0,0 +1,955 @@
+    model: Model
+    _server: _triton_bindings.TRITONSERVER_Server
+    request_id: Optional[str] = None
+    parameters: dict = field(default_factory=dict)
+    outputs: dict = field(default_factory=dict)
+    error: Optional[_triton_bindings.TritonError] = None
+    classification_label: Optional[str] = None
+    final: bool = False
+
+    @staticmethod
+    def _set_from_server_response(
+        model: Model,
+        server: _triton_bindings.TRITONSERVER_Server,
+        request: _triton_bindings.TRITONSERVER_InferenceRequest,
+        response,
+        flags: _triton_bindings.TRITONSERVER_ResponseCompleteFlag,
+        output_array_module: Optional[ModuleType] = None,
+    ):
+        values: dict = {
+            "_server": server,
+            "model": model,
+            "request_id": request.id,
+            "final": flags == _triton_bindings.TRITONSERVER_ResponseCompleteFlag.FINAL,
+        }
+
+        if response is None:
+            return InferenceResponse(**values)
+
+        try:
+            response.throw_if_response_error()
+        except _triton_bindings.TritonError as error:
+            values["error"] = error
+
+        name, version = response.model
+        values["model"] = Model(server, name, version)
+        values["request_id"] = response.id
+        parameters = {}
+        for parameter_index in range(response.parameter_count):
+            name, type_, value = response.parameter(parameter_index)
+            parameters[name] = value
+        values["parameters"] = parameters
+        outputs = {}
+        for output_index in range(response.output_count):
+            (
+                name,
+                data_type,
+                shape,
+                data_ptr,
+                byte_size,
+                memory_type,
+                memory_type_id,
+                memory_buffer,
+            ) = response.output(output_index)
+            tensor = _datautils.Tensor.from_memory_buffer(
+                data_type, shape, memory_buffer
+            )
+            if output_array_module:
+                outputs[name] = tensor.to_ndarray(output_array_module)
+            else:
+                outputs[name] = tensor
+        values["outputs"] = outputs
+
+        # values["classification_label"] = response.output_classification_label()
+
+        return InferenceResponse(**values)
+
+
+@dataclass
+class InferenceRequest:
+    model: Model
+    _server: _triton_bindings.TRITONSERVER_Server
+    request_id: Optional[str] = None
+    flags: int = 0
+    correlation_id: Optional[int | str] = None
+    priority: int = 0
+    timeout: int = 0
+    inputs: dict[str, _datautils.Tensor | Any] = field(default_factory=dict)
+    parameters: dict[str, str | int | bool] = field(default_factory=dict)
+    output_memory_type: Optional[_datautils.DeviceOrMemoryType] = None
+    output_memory_allocator: Optional[_datautils.MemoryAllocator] = None
+    output_array_module: Optional[ModuleType] = None
```

Comment:
Can you explain this array_module stuff being passed around? I saw it in the rayserve example and it feels like an antipattern.

If we need some kind of info at runtime to properly convert, can't we just check the object?
```python
if isinstance(array, np.ndarray):
  ...
elif isinstance(array, cupy.ndarray):
  ...
```

Is it only needed for dlpack logic?

Reply:
I also figured some of the awareness would be handled in the Allocators as well. Are these disjoint?

Reply:
the design here is to allow the user to select the framework they want to use for output (which may be different or the same as input). They also have the option of just using the opaque Tensor object and using dlpack.

So its mostly a convenience - the idea is kind of borrowed from the python array api where you can access the array module from an object like you can get the __dl_pack__ to understand how to transfer the memory.

I think of it as, "output_memory_type" is to select the device

"output_array_module" is to select the framework to use the tensor.

in my head there are three distinct but related concepts:

1) device where memory is allocated
2) allocation framework
3) usage framework

these map to: output_memory_type, output_memory_allocator, and output_array_module

Reply:
> I also figured some of the awareness would be handled in the Allocators as well. Are these disjoint?

Yes - disjoint. To keep things like RMM as a potential - memory is allocated independent of type and shape - just size and device.

Reply:
I feel it would be more intuitive for the user to explicitly control the usage framework portion you described, assuming we provide the necessary interface / conversion magic methods like `__dlpack__` to make it easy for them:
```python
    responses = model.infer(
        inputs={
            "prompt": np.array(
                ["boy with balloon, gritty, urban, charcoal sketch"], dtype="object"
            ).reshape(-1, 1)
        },
        output_memory_type="GPU",
        # output_array_module=torch
    )

    for response in responses:
        output_tensor = response.outputs["generated_image"]
        # generated_image = output_tensor.squeeze().type(torch.uint8)
        generated_image = torch.from_dlpack(output_tensor).squeeze().type(torch.uint8)
```

Reply:
that's exactly how it works! if you don't specify an output_array_module - you get a Tensor (see example #1)

Reply:
I added 'output array module' for completeness in my head -but we can easily remove it and force user to use the tensor object. Just thought it would be nice to have "torch in" "torch out" usability, as opposed to only have "torch in", "tensor out" (but that is the default)

Reply:
Right - I just don't think we should provide another option with output_array_module that passes a literal module around. I think that's an unnecessary extra step for the first iteration. I think our job is just to provide the necessary interfaces to make the Tensor object easy to use/interop - but not to handle the one-liner conversion call for them with an extra module arg. I think it hurts readability/intuition more than helps.

Reply:
The fact that we can call `torch.from_dlpack(Tensor)` or `Tensor.to_ndarray()` etc. on their own looks great to me and I think that's sufficient.

Reply:
sounds good - I think __dlpack__ and to_ndarray are sufficient. 

to_ndarray will still be needed because of BYTES serialization / deserialization - but fallback is __dlpack__ - can remove the output_array_module